src/cascadia/TerminalSettingsModel/Profile.cpp

```diff
@@ -227,40 +221,34 @@ winrt::com_ptr<winrt::Microsoft::Terminal::Settings::Model::implementation::Prof
 // - true iff the json object has the same `GUID` as we do.
 bool Profile::ShouldBeLayered(const Json::Value& json) const
 {
-    // guid was not set
-    const auto myGuid{ Guid() };
-    if (myGuid == winrt::guid{})
-    {
-        return false;
-    }
-
     // First, check that GUIDs match. This is easy. If they don't match, they
     // should _definitely_ not layer.
-    if (const auto otherGuid{ JsonUtils::GetValueForKey<std::optional<winrt::guid>>(json, GuidKey) })
+    const auto otherGuid{ JsonUtils::GetValueForKey<std::optional<winrt::guid>>(json, GuidKey) };
+    const auto otherSource{ JsonUtils::GetValueForKey<std::optional<winrt::hstring>>(json, SourceKey) };
+    if (otherGuid)
     {
-        if (otherGuid.value() != myGuid)
+        if (otherGuid.value() != Guid())
         {
             return false;
         }
     }
     else
     {
-        // If the other json object didn't have a GUID, we definitely don't want
-        // to layer. We technically might have the same name, and would
-        // auto-generate the same guid, but they should be treated as different
-        // profiles.
-        return false;
+        // If the other json object didn't have a GUID,
+        // check if we auto-generate the same guid using the name and source.
+        const auto otherName{ JsonUtils::GetValueForKey<std::optional<winrt::hstring>>(json, NameKey) };
+        if (Guid() != _GenerateGuidForProfile(otherName ? *otherName : L"Default", otherSource ? *otherSource : L""))
+        {
+            return false;
+        }
     }
 
-    std::optional<std::wstring> otherSource;
-    bool otherHadSource = JsonUtils::GetValueForKey(json, SourceKey, otherSource);
-
     // For profiles with a `source`, also check the `source` property.
     bool sourceMatches = false;
     const auto mySource{ Source() };
     if (!mySource.empty())
     {
-        if (otherHadSource)

```

Comment:
there's a very subtle difference between what you have now and what you had before regarding the treatment of `null` in the source field in json.

also, i think it would have resulted in slightly undefined behavior if you had previously had a `null` source in json.